libs/native/src/main/java/org/elasticsearch/nativeaccess/lib/NativeLibraryProvider.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.nativeaccess.lib;
+
+import org.elasticsearch.core.internal.provider.ProviderLocator;
+
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Supplier;
+
+/**
+ * Allows loading native library mappings.
+ */
+public abstract class NativeLibraryProvider {
+
+    private final String name;
+    private final Map<Class<?>, Supplier<?>> libraries;
+
+    protected NativeLibraryProvider(String name, Map<Class<?>, Supplier<?>> libraries) {
+        this.name = name;
+        this.libraries = libraries;
+    }
+
+    /**
+     * Get the one and only instance of {@link NativeLibraryProvider} that is specific to the running JDK version.
+     */
+    public static NativeLibraryProvider instance() {
+        return Holder.INSTANCE;
+    }
+
+    /** Returns a human-understandable name for this provider */
+    public String getName() {
+        return name;
+    }
+
+    /**
+     * Construct an instance of the given library class.
+     * @param cls The library class to create
+     * @return An instance of the class
+     */
+    public <T> T getLibrary(Class<T> cls) {
```

Comment:
I wonder if it's worth introducing a marker interface here for native libraries so we can restrict this down a bit and provide a bit better discoverability for developers. As it is I can pass any class to this method and I have no single place to see what all native access interfaces exist.

We could take this a step further and make said marker interface sealed, there by formalizing all the known sets of native libraries we support. We could then at initialization enforce this, by ensuring all implementations of `NativeLibraryProvider` register implementations for all known subinterfaces of this marker interface. This would allow us to blow up earlier, and with a better error than hitting the assertions below.

Reply:
I don't think we can seal such an interface since the implementations are spread out across 2 modules? The only allowed lib classes are those in this same package, and the intention is to make that only available to the jna module (though currently we must have an unqualified export due to a bug in ProviderLocator).

So without a sealed interface, and given there are only 2 implementations, is a marker interface actually useful?

Reply:
> I don't think we can seal such an interface since the implementations are spread out across 2 modules? 

The _implementations_ wouldn't, only the subintefaces, which themselves could be flagged `non-sealed`. I'm envisioning something like this:

```java
sealed interface NativeLibrary permits PosixCLibrary, ZstdLibrary { }

non-sealed Interface PosixCLibrary { }

non-sealed Interface ZstdLibrary { }
```

The intention here is not to enumerate the _implementations_ but the different _libraries_. Alternatively we just merge all native functions into a single interface and just use that. As it is now there is not formal definition of what interfaces are "native" and should be loaded via this mechanism, or what interfaces a native implementation needs to support. You can dig through the code to find out, but discoverability isn't great.